Add Soul Fire'd and use the provided API.

Crystal-Spider opened the pull request:
## PR to integrate Soul Fire'd API.

Quoting the original comment from #443:

> [Soul Fire'd](https://www.curseforge.com/minecraft/mc-mods/soul-fired) is a mod of mine with a dual aim:
> 
> Add actually unique properties to Soul Fire.
> Provide an easy-to-use API for other mods to register their own, or even of other mods, custom fires and having everything automatically handled.
> I saw that Infernal Expansion can benefit in both cases as it currently is:
> 
> Having its own small API to register some specific fires, either its owns, of other mods, or of Minecraft.
> Handling fires visual effects and some other fire-related stuff.
> By integrating the Soul Fire'd API, Infernal Expansion gets more streamlined code, removal of some mixins, easier fire handling, and more optional features (e.g. enchantments handling, easy fire source and campfire handling, etc.).
> Furthermore, Vanilla fire and soul fire are already handled without needing explicit references in your code.
> I also plan to expand the ecosystem with further PRs to other mods, such that they are all compatible with each other and there's less need for you to register others' fires.
> 
> If this PR is accepted, I will also start working for 1.16.5 and 1.18.2 PRs (or for whichever branch is better, maybe release-3.0?).
> 
> Lastly, I know it's not always the best to add dependency to other mods as that might slow down future versions rollouts, but I can assure you, as the Soul Fire'd API developer, that it's always going to come out very soon for newer versions (and you can already see it by the wide range of Minecraft versions supported and actively mantained).
> 
> Let me know if I did anything wrong when contributing or if something is not clear, or really anything in general.

## Latest news

BYG already accepted a PR to integrate Soul Fire'd (as an optional dependency). This means that:

- SF ecosystem is indeed expanding.
- BYG fire handling won't be explicitly needed anymore: just requiring SF as a dependency of IE will ensure that, when BYG is installed, its fires are handled too.

As for BYG, IE could too add SF as an optional dependency (if you want, I can open another PR).  
But, if you so prefer, it can be added as an optional dependency. This would mean losing most of the streamlined code that comes from this PR. It would also mean that the "fire consistency" feature would be lacking unless SF is present.  
For this reasons, I think it makes sense to add it as a required dependency, but of course it's up to you and I'm fine either way.

## Notes

Soul Fire'd supports, currently, 1.21.4, 1.21.3, 1.21.1, 1.21, 1.20.4, 1.20.2, 1.20.1, 1.19.4, 1.19.2, and 1.18.2.  
If you'd ever need it for another Minecraft version, I can easily port it.  
And I'd also always give full priority support if you ever face issues or have any question about the API.

If this PR is accepted, I might also open another one for Endergetic Expansion so that the part of the code that handles its fire can be removed, as it was the case for BYG.  
I can also, as stated in the original PR, work on other PRs for any other version/branch you want, to make it easy for you to integrate SF.

<details>
  <summary>Choices I made in the changes with this PR that are not strictly related with adding SF</summary>

1. I commented out the line of code generating smoke particles for the wall torch version of the glow torch.  
   This because I saw that the standing torch version does the same.
2. I kept the `GlowFireBlock` class, although not necessary anymore thanks to SF API, because I thought that maybe you'll want to override the `animateTick` method to remove smoke particles and add glow sparkle ones.  
   If this is not something you'll want to do, just tell me and I'll remove the `GlowFireBlock` class that'd be useless at that point.
3. I noticed, testing the latest official released IE version for 1.19.2, that players can't catch glow fire. By this I mean that if the player is in the source block, it does take damage and all that, but I could never manage to keep the fire lasting on the player after leaving the fire source. However, mobs can catch fire as usual. If this was a bug, it was automatically fixed by using Soul Fire'd. If it was not a bug, let me know and I'll restore the feature.
4. Supposing point 3 was a bug, I also moved the logic that gives the player the Luminous effect from the source block to the fire itself. This means that the effect will apply whenever a living entity takes damage due to burning from glow fire, rather than only when inside the source block. If you don't like this, I can easily restore how it was before.
</details>

## Changes summary

<details>
  <summary><h3>Changes</h3></summary>

- `build.gradle`, `build.properties`, and `mods.toml`:  
  Added Soul Fire'd dependency.
- `magma_cube_avoid_blocks.json` and `blockstates/glow_wall_torch.json`:  
  Renamed occurrences of `"infernalexp:glow_torch_wall"` to `"infernalexp:glow_wall_torch"`.
- `infernal-expansion.mixins.json`:  
  Removed lines that referenced removed mixins.
- `InfernalExpansion.java`:  
  Moved the registration of fire types at the top rather than in the common setup.
- `IEParticlesTypes.java`, `IEItems.java`, and `IEBlocks.java`:  
  Replaced objects registration (of fire related ones) with the more streamlined and eased registration via SF API.
- `IEBlockEntityTypes.java`:  
  Removed `GlowCampfireBlockEntity` registration (not needed with SF API).
- `IEFireTypes.java`:  
  Replaced the content with logic to register the fire types to SF API.  
  `GLOW_FIRE_TYPE` was used to register glow fire, its behavior, and the fire related game objects (referenced also in the IE classes for game objects initialization. Enchantments (glow fire aspect and glow flame) are not included, but if you want they can be easily added.  
  `ENDER_FIRE_TYPE` is used to register, if Endergetic Expansion i loaded, the ender fire.  
  Fire types for BYG were removed since BYG already supports SF API.
- `ClientEvents.java`:  
  Removed the registration of the block entity renderer for the glow campfire block entity since it's handled automatically by SF.  
  Removed registration to the cutout render layer for fire related game objects since it's handled automatically by SF.
- `GlowCampfireBlock.java`, `GlowFireBlock.java`, `GlowTorchBlock.java`, and `GlowWallTorchBlock.java`:  
  Changes explained in the Notes section above, along with code streamlining and minor changes to conform to SF API.
</details>
<details>
  <summary><h3>Removed</h3></summary>

- `MixinEntity.java`
- `MixinAbstractFireBlock.java`
- `MixinOverlayRenderer.java`
- `MixinModelBakery.java`
- `MixinEntityRendererManager.java`
- `ClientFireType.java`
- `GlowCampfireBlockEntity.java`
- `FireType.java`
- `FireTypeAccess.java`
</details>
<details>
  <summary><h3>Renamed</h3></summary>

- `blockstates/glow_torch_wall.json` into `blockstates/glow_wall_torch.json`.  
  This is not strictly necessary, as you can specify with SF API any custom name for your fire related objects, but it's better to use the standard name.
- `models/block/glow_torch_wall.json` into `models/block/glow_wall_torch.json`.  
  This is not strictly necessary, as you can specify with SF API any custom name for your fire related objects, but it's better to use the standard name.
</details>

Crystal-Spider committed:
Add Soul Fire'd and use the provided API.

Jonathing commented:
Ok, so there are a lot of breaking changes in this PR, so I don't see myself pulling this in for the next patch. Specifically, IE's FireType has been public API for several years now so I don't want to hastily pull this in.

Another thing is that I am not too fond of the idea of IE having a required dependency on Soul Fire'd (or any mod for that matter). When I worked on [The Midnight](https://www.curseforge.com/minecraft/mc-mods/the-midnight), I solved this using Forge's Jar-in-Jar so users don't have to download and install multiple dependencies manually. So, here are two solutions for this:

1. Infernal Expansion includes Soul Fire'd as an embedded dependency so that users do not have to download it. This comes with the obvious downside that your mod's download counter doesn't go up, but it will still be listed on CurseForge as am embedded dependency of IE.
2. Soul Fire'd is used as an optional dependency of Infernal Expansion and IE uses its own system if Soul Fire'd is not loaded. My plan for this was to rip out IE's current implementation in 3.0.0 and replace it using a system similar to what I made in The Midnight, where an entity capability is used to detect if it is taking damage from our specific flame.

So yeah, it's not as simple as "monkey see, monkey pull" as I'd like it to be, but that's where we are at at the moment. Soul Fire'd is GPLv3 so I can technically include it as is but obviously since people are anal about mod usage and downloads I wanted to make this clear here.

Jonathing commented on src/main/java/org/infernalstudios/infernalexp/InfernalExpansion.java:
```diff
@@ -83,6 +83,8 @@ public class InfernalExpansion {
     public static final String MOD_ID = "infernalexp";
 
     public InfernalExpansion() {
+        IEFireTypes.register();
+
```
Please move this to after the mod loading context and event bus are assigned.

Jonathing commented on src/main/java/org/infernalstudios/infernalexp/blocks/GlowCampfireBlock.java:
```diff
@@ -16,70 +16,47 @@
-            for(int i = 0; i < randomsource.nextInt(2) + 2; ++i) {
-                makeParticles(p_155319_, p_155320_);
+    protected BlockEntityTicker<CampfireBlockEntity> particleTick() {
+        return (Level p_155319_, BlockPos p_155320_, BlockState p_155321_, CampfireBlockEntity p_155322_) -> {
```
I would appreciate using named parameters.

Crystal-Spider replied:
Yeah, I would have, but I just copy-pasted from what was already there. I moved the previous `particleTick` body inside the lambda, keeping everything as it was before. Should I add named parameters now, or would that be considered as something not related to this PR?

Jonathing replied:
Go ahead and add the parameters. Since it's something that this PR changes you might as well. My comment earlier was regarding things that aren't relevant to the PR itself.

Jonathing reviewed (changes requested):
This PR includes a lot of miscellaneous and fixes that aren't related to what you're trying to accomplish. If you find miscellaneous bugs in the code, please don't include them in the PR like this. Instead let me know by opening an issue or making a separate PR. I say this primarily because I'm going to be doing a very large cleanup sweep of the entire code-base. Also, additional cleanup makes PRs like these harder to review, since it can drag away focus from I'm supposed to understand.

Crystal-Spider commented:
> This PR includes a lot of miscellaneous and fixes that aren't related to what you're trying to accomplish. If you find miscellaneous bugs in the code, please don't include them in the PR like this. Instead let me know by opening an issue or making a separate PR. I say this primarily because I'm going to be doing a very large cleanup sweep of the entire code-base. Also, additional cleanup makes PRs like these harder to review, since it can drag away focus from I'm supposed to understand.

I guess you're referring to those 4 points in the Notes section, right?

For 1, I can revert the change no problem.  
For 2, there's not really anything to revert. Either I delete the block class (because it's currently useless), or I leave it there as it is now. I thought of the second to make it easier if you'll need to override the particle behavior, but if I should be strict with the changes I should just delete it. Let me know.  
For 3, it's not really something I did on purpose. It was a bug that was fixed automatically by introducing Soul Fire'd.  
For 4, I thought it'd be something related to the PR since it uses the API to insert this custom behavior (giving the luminous effect), at least to show that it's something possible. Should I move it back to the `entityInside` method of the `GlowFireBlock` class? (which of course changes also what I should do regarding point 2).

If there's anything else you meant by miscellaneous and fixes, let me know!

SwanX1 reviewed (changes requested):
If this PR is to add Soul Fire'd's API for the IE one, I'd like this PR to stick to that, not general code fixes and changes, as @Jonathing has already mentioned. Please revert those changes.
Soul Fire'd should either be an embedded dependency or an optional one. Adding a hard dependency is not favorable for IE.

Crystal-Spider commented:
> If this PR is to add Soul Fire'd's API for the IE one, I'd like this PR to stick to that, not general code fixes and changes, as @Jonathing has already mentioned. Please revert those changes. Soul Fire'd should either be an embedded dependency or an optional one. Adding a hard dependency is not favorable for IE.

Yes, of course. I just need to understand which general code fixes and changes you are talking about, other than the ones in mentioned in my previous reply. But for point 2 and 4, I need to know what you prefer, as there is no clear change to revert.

Crystal-Spider commented:
> Ok, so there are a lot of breaking changes in this PR, so I don't see myself pulling this in for the next patch. Specifically, IE's FireType has been public API for several years now so I don't want to hastily pull this in.
> 
> Another thing is that I am not too fond of the idea of IE having a required dependency on Soul Fire'd (or any mod for that matter). When I worked on [The Midnight](https://www.curseforge.com/minecraft/mc-mods/the-midnight), I solved this using Forge's Jar-in-Jar so users don't have to download and install multiple dependencies manually. So, here are two solutions for this:
> 
> 1. Infernal Expansion includes Soul Fire'd as an embedded dependency so that users do not have to download it. This comes with the obvious downside that your mod's download counter doesn't go up, but it will still be listed on CurseForge as am embedded dependency of IE.
> 2. Soul Fire'd is used as an optional dependency of Infernal Expansion and IE uses its own system if Soul Fire'd is not loaded. My plan for this was to rip out IE's current implementation in 3.0.0 and replace it using a system similar to what I made in The Midnight, where an entity capability is used to detect if it is taking damage from our specific flame.
> 
> So yeah, it's not as simple as "monkey see, monkey pull" as I'd like it to be, but that's where we are at at the moment. Soul Fire'd is GPLv3 so I can technically include it as is but obviously since people are anal about mod usage and downloads I wanted to make this clear here.

I totally understand.  
First of all, thank you for openly talking about the jar-in-jar, although it's something you could legally do.  
Btw, what would happen if IE had SF as jar-in-jar, but also SF was installed separately?

Anyway, my point of view is the following:  
What I'd really like is to expand the Soul Fire'd ecosystem, I'm not really after downloads (although, of course, they do profit me). So I wouldn't be against the jar-in-jar option, although I hope that SF could still be installed separately (for example, in modpacks).  
So, a solution, if possible, could be to include SF as jar-in-jar, but write a little section (even just a sentence) in the mod page. Something along the line of "[Soul Fire'd](https://www.curseforge.com/minecraft/mc-mods/soul-fire-d) is included to handle fire behavior. If you're making a modpack, please include it separately". I mean, I'm not the best at this, so it could be written as you see fit, but the meaning should be something like that.

Btw, on one hand, I don't know how many people still download mods manually one by one rather than using an app like CurseForge, Modrinth, PrismLauncher, etc. that automatically downloads also any dependency. And, even when they do, the mod loader tells exactly what's missing.  
However, on the other hand, I don't like adding hard dependencies to my mods either, unless strictly necessary, so I perfectly understand even considering the above ahah  
The solution mentioned above would be fine for me, if it's fine for you too.

P.S. sorry for my long messages, brevity is not one of my strengths generally, and even more so when it's not my first language 😅

Crystal-Spider committed:
Comply to reviews.

SwanX1 commented:
> Yes, of course. I just need to understand which general code fixes and changes you are talking about, other than the ones in mentioned in my previous reply. But for point 2 and 4, I need to know what you prefer, as there is no clear change to revert.

I would like you to revert all changes not necessary for the usage of the Soul Fire'd API, that is: 
- `assets` and `data` directories (renaming the RL for `infernalexp:glow_torch_wall`
- Any unrelated changes to code, like:
  ```patch
   private static Boolean neverAllowSpawn(BlockState state, BlockGetter reader, BlockPos pos, EntityType<?> entity) {
  -        return (boolean) false;
  +        return false;
   }
  ```

Crystal-Spider committed:
Comply to reviews.

Crystal-Spider committed:
Fix some indentation.

Crystal-Spider commented:
I reverted the torch wall renaming, but I had to add [this line](https://github.com/Crystal-Nest/Infernal-Expansion/blob/0ddcb0ac9ea6451196e8dc3f0ebd3d87a487c07e/src/main/java/org/infernalstudios/infernalexp/init/IEFireTypes.java#L39), which, in my opinion, makes it a little less elegant. That's why I had renamed to use the default value for that component instead.  
Not that it matters, but I thought of point out the reason why I did that, so that later on you can choose and you know what's going on.

Some other changes, like the removing the boolean cast, or changing a list of block imports to import *, were due to my IDE configuration, sorry. I'll try to pay more attention to these details. If you have an IntelliJ IDEA config file I can use for this project, I'll be happy to comply.

Jonathing commented:
> Btw, what would happen if IE had SF as jar-in-jar, but also SF was installed separately?

Forge's Jar-in-Jar system, while scuffed, does work as intended. So, if a mod has another mod installed and the mod already exists in the mods folder, Forge's dependency resolver will choose to load the one with the higher version. Dependency constraints are also taken into account (albeit not that well in its current implementation). So this is a non-issue.

> What I'd really like is to expand the Soul Fire'd ecosystem, I'm not really after downloads (although, of course, they do profit me). So I wouldn't be against the jar-in-jar option, although I hope that SF could still be installed separately (for example, in modpacks).

That would be up to the modpack creator. We are still going to be adament about not having IE depend on an external mod, since it is a huge inconvenience for people who do not use fancy launchers or run pre-made modpacks.

> So, a solution, if possible, could be to include SF as jar-in-jar, but write a little section (even just a sentence) in the mod page. Something along the line of "[Soul Fire'd](https://www.curseforge.com/minecraft/mc-mods/soul-fire-d) is included to handle fire behavior. If you're making a modpack, please include it separately". I mean, I'm not the best at this, so it could be written as you see fit, but the meaning should be something like that.

If we decide to move forward with Soul Fire'd being an embedded dependency, then yes we would make an additional sentence/paragraph/section that mentions it. I obviously don't own Infernal Expansion and am not a major developer of Infernal Studios, so I can't make any promises, but I agree that a section telling modpackers to include it separately would be good. It would also benefit performance at load time since Forge wouldn't have to read from a jar file inside of a jar file (it does not do any extracting).

I already do something similar with The Midnight, where all embedded dependencies are listed on each version uploaded. The jar file itself would also contain a LICENSES.md in the `META-INF/jarjar` folder that shows the GPLv3 license since we would include Soul Fire'd. And yes, we would include a small sentence or something similar in the description regarding its inclusion.

All of this would personally be preferable to making it an optional dependency, as I do not agree with a lot of the current implementation and would prefer to remove it as you have already done so in this PR. And it would *certainly* be better than making it a required dependency for others to download.

> Btw, on one hand, I don't know how many people still download mods manually one by one rather than using an app like CurseForge, Modrinth, PrismLauncher, etc. that automatically downloads also any dependency. And, even when they do, the mod loader tells exactly what's missing.
However, on the other hand, I don't like adding hard dependencies to my mods either, unless strictly necessary, so I perfectly understand even considering the above ahah
The solution mentioned above would be fine for me, if it's fine for you too.

In my opinion, the argument that "the launcher downloads it for you" is invalid because not everyone uses Prism Launcher, CurseForge, or Modrinth. As a Forge developer, I can assure you that we constantly get support inquiries from people who still use the vanilla Minecraft Launcher, even on versions as old as 1.19.2.

SwanX1 commented:
> That would be up to the modpack creator. We are still going to be adament about not having IE depend on an external mod, since it is a huge inconvenience for people who do not use fancy launchers or run pre-made modpacks.
I'll add to this, since most of the time I test a mod outside of the development environment by simply chucking the built mod in the mods folder. Adding a hard installable dependency would hinder that...

Jonathing commented:
> I'll add to this, since most of the time I test a mod outside of the development environment by simply chucking the built mod in the mods folder. Adding a hard installable dependency would hinder that...

I agree. If Soul Fire'd is included as an embedded dependency, if the modpack creator forgets to bump the version, Forge will just use the newer version included in IE's jar.

SwanX1 commented:
> I obviously don't own Infernal Expansion and am not a major developer of Infernal Studios, so I can't make any promises [..]

I can! If we decide to move forward with the embedded dependency, a disclaimer would definitely be added to the description. I have been wanting a "compatability" section in the description for a while now, so if we add it as an optional dependency, that would also be added under the section.

Jonathing commented:
I personally would prefer to see it as an embedded dependency, as that would mean less code for me to maintain. You can decide what you want to do for your 1.20.1 rewrite though. I ain't touching that shit LMAO

Crystal-Spider commented:
Then for me it's fine to proceed with the jar-in-jar!  
To be clear, Soul Fire'd requires [Cobweb](https://www.curseforge.com/minecraft/mc-mods/cobweb), which you can add as a jar-in-jar too of course.  
Until 1.21, it also needs [FCAP](https://www.curseforge.com/minecraft/mc-mods/forge-config-api-port-fabric), but given the [license](https://github.com/Fuzss/forgeconfigapiport/blob/main/LICENSE.md) of that project, it should be fine to add as jar-in-jar too.

Jonathing commented:
Ok now that we're done talking about semantics and dependencies, I'll go ahead and actually test and review the code now.

Crystal-Spider commented:
> Looks like these are GPLv3 and Mozilla so we should be fine on that front.

Yeah, also Cobweb is mine too, so there's definitely no problem about that.

SwanX1 commented:
> I personally would prefer to see it as an embedded dependency, [..]

I'm leaning more on the optional dependency route, as it's lighter on the download size, but that'd be up to @nekomaster1000. Either way, there is a config library (still in the works, somehow), which would be added as an embedded dependency, and an existing gradle config for that would make that less of a hassle (offload the workload to other PRs hehe)

> [..] I ain't touching that shit LMAO

Not nice :(

Jonathing commented:
> I'm leaning more on the optional dependency route, as it's lighter on the download size

It's not something that Infernal Studios would have to worry about unless you host your own projects in your own servers or maven repositories. But yeah I do plan on cleaning up the gradle config to make this kind of work a lot more friendly, since right now it's a bit of a mismatch of multiple different MDKs and plugins (totally understandable btw)

> Not nice :(

I'm evil >:)

SwanX1 commented:
> It's not something that Infernal Studios would have to worry about unless you host your own projects in your own servers or maven repositories.

[We do](https://maven.infernalstudios.org), it's just not something we advertise, and the GitHub workflows have been broken for that for a while.

Jonathing commented:
The council has decided.

![the council](https://github.com/user-attachments/assets/ed5a78ad-5683-4693-b9b0-d653485a7287)

Crystal-Spider commented:
If file size is a concern, I can actually get Soul Fire'd jar from `~250KB` down to `~160KB`, and Cobweb from `~400KB` down to `~80KB`.  
I don't have any control over FCAP, which amounts to `~400KB` (goes down to `~250KB` for 1.20.2 and to `~90KB` for 1.20.4), however it won't be needed at all since 1.21.  
So it'd be very small increase in file size.

I also saw a surge in daily downloads for Soul Fire'd, going from an average of 5k in the past few months, to an average of 10k since the start of January. And BYG already defaults to supporting it, and maybe Endergetic Expansion will follow along.  
And not only mods that add their own custom fires, but also mods like [Tetra](https://www.curseforge.com/minecraft/mc-mods/tetra), [Decorative Blocks](https://www.curseforge.com/minecraft/mc-mods/decorative-blocks), [Danger Close](https://www.curseforge.com/minecraft/mc-mods/danger-close), [Alex's Mobs Interactions](https://www.curseforge.com/minecraft/mc-mods/alexs-mobs-interaction), and [Amendments](https://www.curseforge.com/minecraft/mc-mods/amendments) all support Soul Fire'd, with more to come.
So I really believe in the expanding ecosystem.

Furthermore, as already said, using SF reduces quite a bit the code you'd have to maintain.  
If update or support is a concern, as I already said, I support a wide range of Minecraft versions.  
If there's a version you need that's not supported, I can port it in a short time.  
If a new Minecraft version comes out and you want to port IE to that, realistically it's going to take much less to port SF, so that it's always available whenever you want to port to newer versions.

So, I hope you reconsider about the optionality, and go instead with the jar-in-jar route.  
Thank you, and sorry again for the long message!

Jonathing commented:
After some more discussions Neko has decided to go ahead with the Jar-in-Jar path for this. Which is great because it means less work for me. If there's anything extra I need to know about condensing the file size as much as possible, please let me know.

![approval](https://github.com/user-attachments/assets/dcb7c158-7101-4256-8daa-5d8a3dc7aa85)

If you can add the jarJar dependency in the buildscript yourself that'd be great. But if not, I can do it myself after we pull this.

Crystal-Spider commented:
Tomorrow I'm going to release the new versions that cut down the file size.  
I never used the jarJar. Although I guess it's likely something fairly easy to learn, I think it'd be better for you to do it since you already know exactly what to do and you can do it following the most appropriate conventions.

Jonathing commented:
Ok

Crystal-Spider committed:
Update SF and Cobweb versions to point to the ones with small file size.

Crystal-Spider commented:
Done

Crystal-Spider commented:
> > I obviously don't own Infernal Expansion and am not a major developer of Infernal Studios, so I can't make any promises [..]
> 
> I can! If we decide to move forward with the embedded dependency, a disclaimer would definitely be added to the description. I have been wanting a "compatability" section in the description for a while now, so if we add it as an optional dependency, that would also be added under the section.

A collaborator of mine suggested that, while IE keeps SF and its deps embedded, you could still declare SF as a required dependency on CurseForge and Modrinth. This way, if a user downloads manually, there's no issue and no need to download extra files. However, if a user uses modded launchers and downloads IE from there, the launcher will automatically pull SF and its dependencies, giving to us too the additional download count.  
What do you think? Is this something you'd be fine with?

Jonathing reviewed (changes requested):
If you can DM Neko on Discord, I would appreciate it. I would rather not get into the politics of including library mods inside of mods. I'm here to review code quality and work on the project, not to bargain with how to include Soul Fire'd. Let's keep the PR discussion about the PR.

Speaking of code quality, I went to look at Soul Fire'd itself and have some concerns regarding mod compatibility. I've made two PRs for you, and until they are merged in some capacity, I will remain hesitant on merging this PR:

- Crystal-Nest/soul-fire-d#61
- Crystal-Nest/soul-fire-d#62

I know we just talked about reducing file size and it seems like the work I put in Crystal-Nest/soul-fire-d#62 might deter from that, but note that a `no-embeds.jar` file is generated as part of the build process, and so we would end up including that jar if Infernal Expansion were to also embed MixinExtras. As I've said in that PR though, the inclusion of MixinExtras is well worth it considering the incompatibilities you've created for yourself with your `@Redirect` injectors.

Crystal-Spider committed:
Update Soul Fire'd version.

Crystal-Spider commented:
> If you can DM Neko on Discord, I would appreciate it. I would rather not get into the politics of including library mods inside of mods. I'm here to review code quality and work on the project, not to bargain with how to include Soul Fire'd. Let's keep the PR discussion about the PR.
> 
> Speaking of code quality, I went to look at Soul Fire'd itself and have some concerns regarding mod compatibility. I've made two PRs for you, and until they are merged in some capacity, I will remain hesitant on merging this PR:
> 
> * [[1.19.2] Use ParticleOptions for custom torches Crystal-Nest/soul-fire-d#61](https://github.com/Crystal-Nest/soul-fire-d/pull/61)
> * [[1.19.2] Use compat-friendly injectors from MixinExtras Crystal-Nest/soul-fire-d#62](https://github.com/Crystal-Nest/soul-fire-d/pull/62)
> 
> I know we just talked about reducing file size and it seems like the work I put in [Crystal-Nest/soul-fire-d#62](https://github.com/Crystal-Nest/soul-fire-d/pull/62) might deter from that, but note that a `no-embeds.jar` file is generated as part of the build process, and so we would end up including that jar if Infernal Expansion were to also embed MixinExtras. As I've said in that PR though, the inclusion of MixinExtras is well worth it considering the incompatibilities you've created for yourself with your `@Redirect` injectors.

I updated the Soul Fire'd version IE points to. However, although the no-embeds jar is on Maven, as you can see [here](https://github.com/Crystal-Nest/soul-fire-d/packages/2169702?version=1.19.2-4.0.10), I don't know how to point specifically to that one. I hope I can leave that to you.

I appreciate a lot all your support and effort, so I want to thank you again for your work!

Jonathing commented:
> I don't know how to point specifically to that one. I hope I can leave that to you.

I don't think IE 2.6 will need to use MixinExtras itself, so that is fine. It's a bug in ForgeGradle 6 that I still don't know how to solve, since the way the jarJar extension is handled is really weird. Don't worry about it, and thank you for working with me on this.

Jonathing commented on build.gradle:
```diff
@@ -96,13 +96,27 @@ repositories {
 
     compileOnly fg.deobf("com.github.glitchfiend:TerraBlender-forge:${config.MINECRAFT_VERSION}-${config.TERRABLENDER_VERSION}")
+
+    implementation fg.deobf("it.crystalnest:cobweb-forge:${config.MINECRAFT_VERSION}-${config.COBWEB_VERSION}")
+    implementation fg.deobf("it.crystalnest:soul-fire-d-forge:${config.MINECRAFT_VERSION}-${config.SOUL_FIRE_D_VERSION}")
```
This needs to be Jar-in-Jar'd. I'll handle this myself after this PR is merged. I'd like to give a bit of retrospection on all of this.

The discussion behind the whole library bundling thing has really worn me out, and there are still uncertainties and disagreements as to how to handle it in the future. I do need to take the blame for bringing about the idea of Jar-in-Jar in the first place because it's something I've done with my own mods in the past, but it's clear that I still don't understand the whole scope of the scenario. My primary argument for embedding Soul-Fire'd instead of marking it as a required dependency was for player convenience, but I've done a rather poor job of thinking about the modder side through most of this PR review, and I'm really sorry about that.

So at the end of the day when IE 2.6 is released, if you are still okay with it, we will include Soul-fire'd with the mod and mark it as required so that launchers can download it as well. If not, then I'm willing to revert that and move forward with at least making it a required dependency for IE. In any case, given the work that we've put into this I don't see myself backing out of IE 1.19.2 using Soul-Fire'd, especially considering how much work it's saved me personally.

Crystal-Spider replied:
Embedded is fine, especially when it's also marked as required for launchers!  
You can proceed, and I'm glad Soul Fire'd was able to save you time!

Jonathing reviewed (approved):
Will merge this later, sometime later today or next week. Thanks again.
